### PR TITLE
Clustering Details

### PR DESCRIPTION
**Pull request by tuncerkaplankiran**
Details for pattterns (clustering) - sampling modes for clustering explained.
auto detect line pattern vs custom pattern  explained.

### CODE REVIEW

File: appendices/clustering.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Clustering Details
+
+When a new event passes through the pipeline, the variants are identified via a proprietary Ragel FSM Based Tokenization process (80%-90% more efficient than regex) and stripped from the event, replaced with wildcards, and the remaining invariant components of the event are then compared to existing pattern sets to determine their similarities.
+
+There are 2 sampling modes for calculating similarities:
```
Comment (taylanisikdemir) [inline appendices/clustering.md]: nit: drop "sampling"
"There are  modes for calculation similarities between patterns"

File: appendices/clustering.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Clustering Details
+
+When a new event passes through the pipeline, the variants are identified via a proprietary Ragel FSM Based Tokenization process (80%-90% more efficient than regex) and stripped from the event, replaced with wildcards, and the remaining invariant components of the event are then compared to existing pattern sets to determine their similarities.
```
Comment (taylanisikdemir) [inline appendices/clustering.md]: nit: "event" might indicate something else. "log" is more appropriate here

File: appendices/clustering.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+There are 2 sampling modes for calculating similarities:
+* [Drain](https://docs.edgedelta.com/appendices/clustering#drain)
+* [Levenshtein Distance](https://docs.edgedelta.com/appendices/clustering#levenshtein-distance)
+
+The goal of calculating similarities is to transform raw log messages into structured log messages by clustering them based on their similarities.
+
+## Drain
+
+Drain is a log parsing algorithm that we used for clustering our logs, it is based on parse tree with fixed depth to guide the log group search process, which effectively avoids constructing a very deep and unbalanced tree.
```
Comment (taylanisikdemir) [inline appendices/clustering.md]: nit: "our logs" -> "logs"

File: appendices/clustering.md (hunk 1 above)
Comment (taylanisikdemir) [reply appendices/clustering.md]: Also mention that "Drain is the default mode"

**Review by taylanisikdemir — APPROVED** [review]
thanks for documenting clustering. with some additional examples this can be turned into a great blog post as well

File: appendices/clustering.md (hunk 1 above)
Comment (tuncerkaplankiran) [reply appendices/clustering.md]: Done!

File: appendices/clustering.md (hunk 2 above)
Comment (tuncerkaplankiran) [reply appendices/clustering.md]: Changed to log.

File: appendices/clustering.md (hunk 3 above)
Comment (tuncerkaplankiran) [reply appendices/clustering.md]: Changed to logs

**Pull request merged**